controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,349 @@
+
+	if err != nil && !strings.Contains(err.Error(), "notFound") {
+		return config, err
+	} else if err == nil {
+		logrus.Debugf("Cluster %v delete is called. Status Code %v", config.Spec.ClusterName, operation.HTTPStatusCode)
+	} else {
+		logrus.Debugf("Cluster %s doesn't exist", config.Spec.ClusterName)
+	}
+
+	return config, nil
+}
+
+func (h *Handler) create(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		locationRRN(config.Spec.ProjectID, config.Spec.Region),
+		generateClusterCreateRequest(config)).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		return config, err
+	}
+
+	if err == nil {
+		logrus.Debugf("Cluster %s create is called for project %s and region/zone %s. Status Code %v",
+			config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region, operation.HTTPStatusCode)
+	}
+
+	config = config.DeepCopy()
+	config.Status.Phase = gkeConfigCreatingPhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) checkAndUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+
+	//TODO: implement -- currently a noop.
+	return config, nil
+}
+
+func (h *Handler) waitForCreationComplete(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	lastMsg := ""
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+		if err != nil {
+			return config, err
+		}
+		cluster, err := svc.Projects.Locations.Clusters.Get(clusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+		if err != nil {
+			return config, err
+		}
+		if cluster.Status == runningStatus {
+			log.Infof(ctx, "Cluster %v is running", config.Spec.ClusterName)
+			config = config.DeepCopy()
+			config.Status.Phase = gkeConfigActivePhase
+			return h.gkeCC.UpdateStatus(config)
+		}
+		if cluster.Status != lastMsg {
+			log.Infof(ctx, "%v cluster %v......", strings.ToLower(cluster.Status), config.Spec.ClusterName)
+			lastMsg = cluster.Status
+		}
+		time.Sleep(time.Second * 5)
+	}
+
+}
+
+func waitClusterRemoveExp(ctx context.Context, svc *gkeapi.Service, config *gkev1.GKEClusterConfig) (*gkeapi.Operation, error) {
+	var operation *gkeapi.Operation
+	var err error
+
+	for i := 1; i < 12; i++ {
+		time.Sleep(time.Duration(i*i) * time.Second)
+		operation, err = svc.Projects.Locations.Clusters.Delete(clusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+		if err == nil {
+			return operation, nil
+		} else if !strings.Contains(err.Error(), "Please wait and try again once it is done") {
+			break
+		}
+	}
+	return operation, err
```

Comment:
Is there a scenario where the `operation` will remain initialized?

Reply:
if you are asking why it is returned here is because we report the status code from it. could maybe be moved from this function, but if another placed used it this might not be true. I thought it best to return it.

Reply:
I am wondering if there's a scenario where `operation` would never be assigned anything (child function returns `nil`).

Reply:
I mean. yes. if `err` is returned operation is `nil`. is that what you are asking? I guess maybe i'm confused.

Reply:
That is what I am asking. I am wondering if it is ever possible for `operation` and  `err` to both be `nil`.